### PR TITLE
Convert proxy factory auto generate mode to integer

### PR DESCRIPTION
**Pull request by dragosprotung**
Clean PR based on #815

### CODE REVIEW

File: lib/Doctrine/Common/Proxy/AbstractProxyFactory.php [hunk 1]
```diff
@@ -69,6 +69,13 @@
      */
     const AUTOGENERATE_EVAL = 3;
 
+    private const AUTOGENERATE_MODES = [
+        self::AUTOGENERATE_NEVER,
+        self::AUTOGENERATE_ALWAYS,
+        self::AUTOGENERATE_FILE_NOT_EXISTS,
+        self::AUTOGENERATE_EVAL
```
Comment (Majkl578) [inline lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: There should be comma after last item.

File: lib/Doctrine/Common/Proxy/AbstractProxyFactory.php [hunk 2]
```diff
@@ -93,12 +100,19 @@
...
+     *
+     * @throws \Doctrine\Common\Proxy\Exception\InvalidArgumentException When auto generate mode is not valid.
      */
     public function __construct(ProxyGenerator $proxyGenerator, ClassMetadataFactory $metadataFactory, $autoGenerate)
     {
         $this->proxyGenerator  = $proxyGenerator;
         $this->metadataFactory = $metadataFactory;
-        $this->autoGenerate    = (bool)$autoGenerate;
+
+        if ( ! in_array($autoGenerate, self::AUTOGENERATE_MODES, false)) {
```
Comment (Majkl578) [inline lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: You don't need false here as it's default behavior.

Comment (dragosprotung) [reply lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: It is the default value, but I wanted to be explicit that we allow other values

Comment (lcobucci) [reply lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: And why not casting it to int and making a strict comparison?

**Comment by lcobucci** [comment]
@dragosprotung there are some CS violations, I've just pushed #817 to update our build process and verify coding standard violations. Do you want me to point out the issues or do you prefer to wait for it to be merged and just use `phpcs/phpcbf`?

File: lib/Doctrine/Common/Proxy/AbstractProxyFactory.php (hunk 2 above)
Comment (dragosprotung) [reply lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: Normally I don't set properties until they are validated. But I guess in this case makes no difference as it's the constructor

**Comment by dragosprotung** [comment]
@lcobucci I can see them from your PR, but I will wait for it to be merged if it will not take too long

**Comment by Ocramius** [comment]
@lcobucci let's not pick on CS if a massive conflicting PR is coming in anyway :-P

File: lib/Doctrine/Common/Proxy/AbstractProxyFactory.php [hunk 3]
```diff
@@ -69,6 +69,13 @@
      */
     const AUTOGENERATE_EVAL = 3;
 
+    private const AUTOGENERATE_MODES = [
```
Comment (Ocramius) [inline lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: YAY!

**Pull request merged**